### PR TITLE
add initial cmake build

### PR DESCRIPTION
**Pull request by rem1776**
adds a cmakelists to build the coupler code and any components. I only used it for the null model, but tried to make things generic enough so that it can be easily used for other models.

This also includes a run test using ctest from the null model build script.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+project(fmscoupler LANGUAGES Fortran)
+
+include(FetchContent)
+
+# define some file locations
+set(EXTERNAL_SOURCE_DIR ${CMAKE_SOURCE_DIR}/external)
+file(MAKE_DIRECTORY ${EXTERNAL_SOURCE_DIR})
+
+# Options for local dependencies
+set(fms_path 			  "${CMAKE_SOURCE_DIR}/../FMS"        CACHE STRING "Path to local fms repository")
```
Comment (mlee03) [inline CMakeLists.txt]: the spacing seems off, not sure if it's intentional

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+set(atmos_path			"${CMAKE_SOURCE_DIR}/../atmos_null" CACHE STRING "Path to local atmospere component repository")
+set(ocean_path			"${CMAKE_SOURCE_DIR}/../ocean_null" CACHE STRING "Path to local ocean component repository")
+set(land_path			  "${CMAKE_SOURCE_DIR}/../land_null"  CACHE STRING "Path to local land component repository")
+set(ice_path			  "${CMAKE_SOURCE_DIR}/../ice_null"   CACHE STRING "Path to local ice component repository")
+set(ice_param_path  "${CMAKE_SOURCE_DIR}/../ice_param"  CACHE STRING "Path to local ice_param component repository")
+
+# Options for cloning dependencies
+option(FETCH_FMS				"Clone FMS repository for building" OFF)
+option(FETCH_COMPONENTS "Clone model component repositories for building" OFF)
+set(fms_tag 			"main" CACHE STRING "Branch/Tag name for FMS. Only used if -DFETCH_FMS=on")
```
Comment (mlee03) [inline CMakeLists.txt]: Instead of main, set the tags to the latest releases?

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	CACHE STRING "Repository URL for ice_param component. Only used if -DFETCH_COMPONENTS=on")
+
+# TODO some more error checking here to ensure fetching/local options make sense
+if(NOT FETCH_FMS)
+  message(NOTICE "-DFETCH_FMS=off - expecting local FMS repository at ${fms_path}.")
+endif()
+
+if(NOT FETCH_COMPONENTS)
+  message(NOTICE
+    "-DFETCH_COMPONENTS=off - expecting components locally available in ${CMAKE_SOURCE_DIR}/.. if path not specified"
```
Comment (mlee03) [inline CMakeLists.txt]: "repositories for each component locally available ... " to be a bit more consistent with the FMS_FETCH message

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    message(STATUS "Cloning ${component_url} to ${EXTERNAL_SOURCE_DIR}/${component_name}")
+    FetchContent_Declare(
+      ${component_name}
+      GIT_REPOSITORY ${component_url}
+      SOURCE_DIR ${EXTERNAL_SOURCE_DIR}/${component_name}
+      GIT_TAG ${${component_name}_tag}
+    )
+    FetchContent_MakeAvailable(${component_name})
+  else()
+    set(component_path ${${component_name}_path} )
```
Comment (mlee03) [inline CMakeLists.txt:105]: The symlink option seems to be overcomplicating things.. is it needed?

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+endif()
+
+if(NOT FETCH_COMPONENTS)
+  message(NOTICE
+    "-DFETCH_COMPONENTS=off - expecting components locally available in ${CMAKE_SOURCE_DIR}/.. if path not specified"
+  )
+endif()
+
+# build FMS, cloning if -DFETCH_FMS=on, otherwise expects a local path provided
+if(FETCH_FMS)
```
Comment (mlee03) [inline CMakeLists.txt]: I can see the use of turning off and on the FETCH_FMS option for FMS testing, but not sure if FETCH_COMPONENTS is necessary.  It makes cmake more complicated than necessary.. Get rid of FETCH_COMPONENTS and always fetch the null components?

Comment (mlee03) [reply CMakeLists.txt]: Maybe also add a comment somewhere about the FMS being a dependency for the other components too.

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+set(SIMPLE_EXEC_SRC ${CMAKE_SOURCE_DIR}/simple/coupler_main.F90)
+set(FULL_EXEC_SRC ${CMAKE_SOURCE_DIR}/full/coupler_main.F90)
+
+# library for simple coupler modules + shared
+add_library(coupler_shared_simple STATIC ${SIMPLE_MOD_SRCS} ${SHARED_MOD_SRCS})
+set_target_properties(coupler_shared_simple PROPERTIES
+    Fortran_MODULE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/simple_coupler_mods"
+)
+set_target_properties(coupler_shared_simple PROPERTIES COMPILE_FLAGS "-fdefault-real-8")
+target_compile_definitions(coupler_shared_simple PUBLIC use_AM3_physics _USE_LEGACY_LAND_)
```
Comment (mlee03) [inline CMakeLists.txt]: the land macro is not used in simple

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+  ice_param
+)
+
+# library for full coupler modules + shared
+add_library(coupler_shared_full STATIC ${FULL_MOD_SRCS} ${SHARED_MOD_SRCS})
+set_target_properties(coupler_shared_full PROPERTIES
+    Fortran_MODULE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/full_coupler_mods"
+)
+set_target_properties(coupler_shared_full PROPERTIES COMPILE_FLAGS "-fdefault-real-8")
+target_compile_definitions(coupler_shared_full PUBLIC use_AM3_physics _USE_LEGACY_LAND_)
```
Comment (mlee03) [inline CMakeLists.txt]: Would it be a major pain to add compilations with and without the macros? 😁

File: README.md [hunk 8]
```diff
@@ -29,6 +29,49 @@ The coupling between the models is designed to conserve fluxes. For coupled mode
 specification file is used to initialize the model grids and perform exchanges between the models.
 The next sections describe how this file and associated grids are used in the coupler.
 
+### Software Structure and Build System
+
+The FMScoupler repository contains 2 distinct "driver" programs, "full" and "simple", which serve as the executable
```
Comment (mlee03) [inline README.md]: The FMSCoupler repository consists of 
1.  `full` subdirectory containing the main driver program and flux exchange modules for fully coupled GFDL models. 
2.  `simple` subdirectory containing the main driver program, a simpler flux exchange module, and a simpler ice model for coupled models only consisting of atmosphere, land, and the simpler ice; 
3.  `shared` subdirectory containing the surface_flux module shared between `full` and `simple`.

File: README.md [hunk 9]
```diff
@@ -29,6 +29,49 @@ The coupling between the models is designed to conserve fluxes. For coupled mode
...
+program to run GFDL developed climate models. It also contains a collection of helper modules that define routines used
+in the drivers, which are split between the shared/ and the full/simple subdirectories.
+
+The "full" coupler has 5 direct dependencies: [FMS](https://github.com/noaa-gfdl/fms), `ice_model_mod`, `land_model_mod`,
```
Comment (mlee03) [inline README.md]: instead of direct, external?

File: README.md [hunk 10]
```diff
@@ -29,6 +29,49 @@ The coupling between the models is designed to conserve fluxes. For coupled mode
...
+`atmos_model_mod`, and `ocean_model_mod`. The "simple" coupler differs in that it does not require an ocean model and
```
Comment (mlee03) [inline README.md]: ... The "simple" coupler does not require `ocean_model_mod` and uses its own internal `ice_model_mod` that depends on the external `ice_param` library.  Thus, the "simple coupler has the following external dependencies:  FMS, `ice_model_mod`, `land_model_mod`, `atmos_model_mod` and `ice_param`

File: README.md [hunk 11]
```diff
@@ -29,6 +29,49 @@ The coupling between the models is designed to conserve fluxes. For coupled mode
...
+includes its own `ice_model` module rather than using an external repository. `ice_param` is a dependency of the simple
+`ice_model`, so it is also included in the build. It is commonly used by GFDL ice models as well, so it will be linked with
+the ice component by default.
+
+The component modules can be provided by a number of different repositories that define the
```
Comment (mlee03) [inline README.md]: Not sure if lines 44-45 are needed.

**Review by mlee03 — COMMENTED** [review]
only suggestions :)

File: CMakeLists.txt (hunk 1 above)
Comment (rem1776) [reply CMakeLists.txt]: vim betrayed me :(

File: CMakeLists.txt (hunk 5 above)
Comment (rem1776) [reply CMakeLists.txt]: Yeah im glad you commented on this because I was kinda torn on how to approach the components. The way I did it has the benefit of being capable of compiling really any model components (in theory...) but in practice, we will be using fre-cli to clone everything so not really needed.

One alternative I thought of was adding an explicit option for the null model build, rather than leaving all these other options around. So I would remove all the FETCH and TAG options and just add a NULL_BUILD or something like that and only do clones if using that one, otherwise it will just compile whats available locally.

File: CMakeLists.txt (hunk 2 above)
Comment (rem1776) [reply CMakeLists.txt]: Since this is used in the CI and the null components get updated somewhat regularly, I think in this case its better to use the main branch

File: CMakeLists.txt (hunk 4 above)
Comment (rem1776) [reply CMakeLists.txt:105]: I might be able to remove that. It's basically moving all the components from outside the coupler to inside a folder called 'external'. Its mainly to make things cleaner from a build perspective.

File: CMakeLists.txt (hunk 1 above)
Comment (mlee03) [reply CMakeLists.txt]: long live emacs!

File: CMakeLists.txt (hunk 5 above)
Comment (mlee03) [reply CMakeLists.txt]: The NULL_BUILD approach sounds better but just worried that if the option is available, we need to test regularly to make sure non-Null components build with this cmake.  I say go with just building a NULL model and add features to build any models if the need arises?

Comment (rem1776) [reply CMakeLists.txt]: Yeah i can clean up the fetching so its just the null components. I think the option can be even more specific, something like FETCH_NULL_MODEL, since the null build can be done with manual cloning too.

File: CMakeLists.txt (hunk 7 above)
Comment (rem1776) [reply CMakeLists.txt]: No its easy, i'll add that in. I think fre will most likely end up setting the macros for normal builds so will be useful.

Comment (rem1776) [reply CMakeLists.txt]: Wait you mean like as a build option right?
